handshake_server.go

```diff
@@ -879,12 +879,19 @@ func (hs *serverHandshakeState) clientHelloInfo() *ClientHelloInfo {
 		pskBinder = hs.clientHello.psks[0].binder
 	}
 
+	var signatureSchemes []SignatureScheme
+	if hs.clientHello.vers > VersionTLS12 && len(hs.clientHello.supportedSignatureAlgorithmsCert) > 0 {
+		signatureSchemes = hs.clientHello.supportedSignatureAlgorithmsCert
+	} else {
+		signatureSchemes = hs.clientHello.supportedSignatureAlgorithms
+	}
+
 	hs.cachedClientHelloInfo = &ClientHelloInfo{
 		CipherSuites:      hs.clientHello.cipherSuites,
 		ServerName:        hs.clientHello.serverName,
 		SupportedCurves:   hs.clientHello.supportedCurves,
 		SupportedPoints:   hs.clientHello.supportedPoints,
-		SignatureSchemes:  hs.clientHello.supportedSignatureAlgorithms,
+		SignatureSchemes:  signatureSchemes,
```

Comment:
I don't think so. SignatureSchemes list is used to get a certificate which has a public key type that client is capable of verifying. For example, client may be able to verify handshake signature that was done with eddsa, but not able to verify certificate for eddsa. Then it will send ``signature_algorithms={eddsa, rsa}`` and ``signature_algorithms_cert = {rsa}``. In such case server must return RSA certificate and must not return EdDSA.
This is only true for TLS 1.3 where this extension is mandatory.

Reply:
Let's say the client has a root certificate C1, who's public key is for a signature scheme not supported by 1.3---let's call it scheme1. The server has a certificate C2 signed by C1, who's public key is for a signature scheme that *is* supported by 1.3---let's call it scheme2. Now, the ClientHello has the following extensions:
* `signature_algorithms` = {scheme2}
* `signature_algorithms_cert` = {scheme1}

So the client is indicating that it's *only* willing to use scheme1 to verify the server's certificate, and it's *only* willing to use scheme2 to verify the handshake. Do you agree that this is valid in 1.3?

Here's the bug. As written, `clientHelloInfo()` returns a `*ClientHelloInfo` with `SignatureSchemes` set to {scheme1}.  Suppose that the server has set the `GetCertificate` parameter of its `Config`. When the server invokes `GetCertificate()`, it will find that the client can verify its certificate; since it supports scheme1, it can verify that C1 signed for C2). However, the client can't verify a signature produced by the secret key of C2. The server will have abort or try to send an HRR.

Do you follow me? Am I missing something?

Reply:
I know what you mean, but in practice those sets are not disjoint, which maybe is not clear from looking at spec.

> Do you agree that this is valid in 1.3?

Well, not really. Such client implementation would be wrong. In particular I disagree with a sentence "Let's say the client has a root certificate C1,". If client "has" such certificate, then it supports algorithm that certificate uses to sign other certificates.  
Please note that it is complatelly fine if this list doesn't contain signing algorithm used by C1 (certificate that begins certification path).


> However, the client can't verify a signature produced by the secret key of C2, since it appears to not support scheme2

Well, server needs to use Scheme2 for TLS handshake signing. It may or may not be possible with public key from C2 (for example client support RSA-PSS for verifying online signatures but not for cert verification. Still RSA public key can be used for both).

> If the signature_algorithms_cert extension is indicated, then I believe your code will only work if signature_algorithms is a subset of signature_algorithms_cert.

I agree. That's how it is implemented in tris.

The goal of this extension is to filter algorithms that can be used for verification. (https://www.ietf.org/mail-archive/web/tls/current/msg25282.html).

Reply:
> Well, not really. Such client implementation would be wrong. In particular I disagree with a sentence "Let's say the client has a root certificate C1,". If client "has" such certificate, then it supports algorithm that certificate uses to sign other certificates.

My point is that it's permissible, according to the spec, for the sets to be disjoint. Do you agree with this statement? If so, then the code is non-compliant as is.

I understand the intent of the extension (and why draft 23 exists at all), and I appreciate that, in practice, `signature_algorithms` could just be a subset of `signature_algorithms_cert`. That's fine if this is *always* the case, but I don't see any reason why this would be guaranteed. What if a boring or OopenSSL client offers a tris server disjoint sets?

But hey, this might not really matter in practice! I'll leave it you to decide if 100% compliance is necessary here.

Reply:
I think the code is both correct and implemented as indicated as in the spec. Yes, I agree spec could be improved. If other client starts offering two disjoint sets - garbage in, garbage out. I prefer this than violating spec by not respecting mandatory extensions (or having weird code trying to deal with garbage).

Reply:
But from my read, there's nothing in the spec that mandates signature_algorithms be a subset of siganture_algorithms_cert.

Reply:
I agree, by the spec they can be disjoint. That's a problem in protocol design.

Reply:
Interesting. Why wouldn't the handshake work? (In the meantime, I'm studying boringSSL to see how they do this.)

Reply:
In BSSL seems to ignore this extension when received. NSS implement it partially also.
Actually I'm not sure what will happen with a handshake, my guess something weird may happen if you send scheme1={ecdsa}, scheme2={rsa}. But I havn't checked

Reply:
I think the handshake should work just fine, as long the set in signature_algorithms_cert is sufficient to verify the certificate, and siganture_algorithms is sufficient to verify the handshake.

Reply:
BSSL appears to process the extension, see ssl/t1_lib.cc:1071 (`static bool ext_sigalgs_cert_add_clienthello(SSL_HANDSHAKE *hs, CBB *out) { ... }`)

Reply:
this is client side, you would need to look at ``tls_extensions`` struct. It has func pointer called ``parse_clienthello`` (t1_lib.cc:576). This pointer is called by the server when parsing CH. t1_lib.cc:3284. Looking at ``kExtensions`` which stores the callbacks for all extensions you can see that ``parse_clienthello`` for `` TLSEXT_TYPE_signature_algorithms_cert`` is set to ``ignore_parse_clienthello`` (t1_lib.cc:2919). 
I like this code actually, good stuff ™

Reply:
Anyways, I think we are good here, isn't?

Reply:
Gotcha, thanks for pointing that out.

Reply:
No, I'm not satisfied that the code is complaint with the spec. But you should feel free to overrule me, because I don't own this code. If I were you, I would run it by Brendan or Nick or whoever else might have a stake in this.

Reply:
Anybody is welcome to review

Reply:
It's non-complaint because it doesn't permit the sets of certificate-signature verification algorithms to be disjoint from the set of handshake-signature verification algorithms. This condition is permitted by the spec, and I believe that the handshake could still be done.

Reply:
It's very confusing I've to say - now you say the code is non-complaint with the spec, or that it is complaint (as you said above)?

Reply:
"No, I'm not satisfied that the code is compliant with the spec." I think before I wrote "complaint" instead of "compliant", which is super confusing! Sorry about that. What I meant is that I don't think the code is compliant.

Reply:
Definitely not the case. The code is fully compliant with the spec. Spec clearly states which list of algorithms must be used for handshake and which must be used for cert validation. Implementation follows the spec in each and every detail. Spec states that implementation MAY use different approach, but current implementation doesn't. Nevertheless, implementation is 100% compliant with TLS 1.3. Indeed, specification of TLS 1.3 is different than 1.2.

Reply:
Well, I remain unconvinced for the reasons I've already mentioned. But I'm happy to drop it.

Reply:
no consensus, let's drop this change.